### PR TITLE
feat: added event emitter to cuba-rest-js for handling fetch errors

### PR DESCRIPTION
**Pull request by abarinovhaulmont**

### CODE REVIEW

File: packages/cuba-rest-js/src/event_emitter.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
+interface IKeyFuncVal {
+    [k: string]: (...args) => any;
+}
+
+class EventEmitter<Events extends IKeyFuncVal, Key extends keyof Events = keyof Events> {
```
Comment (vyacheslav-pushkin) [inline packages/cuba-rest-js/src/event_emitter.ts:5]: I'm not sure we need that much typings complexity. Even for generic case. For example, I can't think of a realistic situation when we need `Events` to be something else than `IKeyFuncVal`. I'd simplify it to make code more readable / easier to follow. Also by doing so we might get rid of `@ts-ignore` below.

Perhaps we could leave it as it is for now but create a separate ticket.

File: packages/cuba-rest-js/src/event_emitter.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    private listeners: Map<Key, Set<Events[Key]>> = new Map();
+
+    emit<K extends Key>(eventName: K, ...restParams: Parameters<Events[K]>) {
```
Comment (vyacheslav-pushkin) [inline packages/cuba-rest-js/src/event_emitter.ts:8]: I have a bunch of ESLint warnings on methods and `REST_EMITTERS_CALLBACKS_PARAMS`. Could you please check if your IDE is using the right config? (`packages/cuba-rest-js/tslint.json`)

File: packages/cuba-rest-js/src/event_emitter.ts (hunk 1 above)
Comment (vyacheslav-pushkin) [reply packages/cuba-rest-js/src/event_emitter.ts:5]: @web-devel WDYT?

File: packages/cuba-rest-js/src/cuba.ts [hunk 3]
```diff
@@ -99,6 +101,11 @@ export interface LoginOptions {
   tokenEndpoint: string;
 }
 
+interface ICubaRestCheckStatusError {
+  message: string;
+  response: Response;
```
Comment (vyacheslav-pushkin) [inline packages/cuba-rest-js/src/cuba.ts:106]: Both should be optional.

File: packages/cuba-rest-js/src/event_emitter.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        const events = this.listeners.get(eventName);
+
+        if (events) events.delete(fn);
+    }
+}
+
+interface ICubaRestCheckStatusError {
+    message: string;
+    response: Response;
+}
```
Comment (vyacheslav-pushkin) [inline packages/cuba-rest-js/src/event_emitter.ts:51]: Same interface in two places.

File: packages/cuba-rest-js/src/event_emitter.ts (hunk 1 above)
Comment (web-devel) [reply packages/cuba-rest-js/src/event_emitter.ts:5]: agree

**Review by web-devel — COMMENTED** [review]
>   chore: added event emitter to cuba-rest-js for handling fetch errors

I think it's `feat` rather then `chore`

**Review by vyacheslav-pushkin — APPROVED** [review]

**Pull request merged**